File: TypeScript/Solution with StringConcatenation.ts

function compressedString(word: string): string {
    let compressed = "";
    let frequency = 1;

    for (let i = 1; i < word.length; ++i) {
        if (frequency === 9 || word.charAt(i - 1) !== word.charAt(i)) {
            compressed += frequency + word.charAt(i - 1);
            frequency = 1;
            continue;
        }
        ++frequency;
    }
    compressed += frequency + word.charAt(word.length - 1);

    return compressed;
};

File: C++/Solution with StringConcatenation.cpp

#include <string>
using namespace std;

class Solution {

public:
    string compressedString(const string& word) const {
        string compressed;
        compressed.reserve(2 * word.size());
        int frequency = 1;

        for (int i = 1; i < word.size(); ++i) {
            if (frequency == 9 || word[i - 1] != word[i]) {
                compressed.push_back(frequency + '0');
                compressed.push_back(word[i - 1]);
                frequency = 1;
                continue;
            }
            ++frequency;
        }
        compressed.push_back(frequency + '0');
        compressed.push_back(word[word.size() - 1]);

        return compressed;
    }
};
